### PR TITLE
feat: add support for L1 to L3 bridging

### PR DESCRIPTION
**Pull request by godzillaba**
gas estimation is probably the most complicated part of this code.

For ERC20 tokens, use `Erc20L1L3Bridger`, for ETH, use `EthL1L3Bridger`

this is how the `Erc20L1L3Bridger` should be used, note that getDepositRequest returns a feetoken amount

also see the tutorials: https://github.com/OffchainLabs/arbitrum-tutorials/pull/117
```typescript
const bridger = new Erc20L1L3Bridger(l3Network)

// our deposit parameters
const depositParams = {
  erc20L1Address: l1Token.address,
  to: l3Recipient,
  amount,
  l1Signer,
  l2Provider: l2Signer.provider!,
  l3Provider,
}

// get a deposit request
const depositRequest = await bridger.getDepositRequest(depositParams)

// approve fee token if necessary
if (depositRequest.feeTokenAmount.gt(0)) {
  await (
    await bridger.approveFeeToken({
      l1Signer,
      l2Provider: l2Signer.provider!,
      amount: depositRequest.feeTokenAmount,
    })
  ).wait()
}

// approve token
await (
  await bridger.approveToken({
    erc20L1Address: l1Token.address,
    l1Signer,
    amount: depositParams.amount,
  })
).wait()

// deposit, use the deposit request we generated earlier
// if the fee token is already approved, you don't really need to getDepositRequest first, you can just pass depositParams to deposit
const depositTx = await bridger.deposit({
  l1Signer,
  txRequest: depositRequest.txRequest,
})

const depositReceipt = await depositTx.wait()

// get deposit status, can pass in txHash, tx, or txReceipt
const status = await bridger.getDepositMessages({
  txReceipt: depositReceipt,
  l1Provider: l1Signer.provider!,
  l2Provider: l2Signer.provider!,
  l3Provider,
})
```

### CODE REVIEW

File: src/lib/assetBridger/l1l3Bridger.ts [hunk 1]
```diff
@@ -0,0 +1,1381 @@
...
+
+    let l1FeeTokenAddress: string | undefined
+
+    try {
+      l1FeeTokenAddress = await this.l2Erc20Bridger.getL1ERC20Address(
+        this.l2FeeTokenAddress,
+        l2Provider
+      )
+    } catch (e: any) {
+      // todo: this feels like a hack
```
Comment (godzillaba) [inline src/lib/assetBridger/l1l3Bridger.ts:413]: this function is pretty weird

File: src/lib/assetBridger/l1l3Bridger.ts [hunk 2]
```diff
@@ -0,0 +1,1382 @@
...
+  l1Signer: Signer
+  overrides?: PayableOverrides
+}
+
+type RetryableGasValues = {
+  gasLimit: BigNumber
+  maxSubmissionFee: BigNumber
+}
+
+export type DepositRequestResult = {
```
Comment (godzillaba) [inline src/lib/assetBridger/l1l3Bridger.ts:86]: should types like this be renamed to something like `Erc20L1L3DepositRequestResult` to remove ambiguity?

File: src/lib/assetBridger/l1l3Bridger.ts [hunk 3]
```diff
@@ -0,0 +1,1382 @@
...
+        amount: BigNumber.from(params.partialTeleportParams.amount),
+      })
+    }
+  }
+
+  /**
+   * Given TeleportParams without the gas parameters, return TeleportParams with gas parameters populated.
+   * Does not modify the input parameters.
+   */
+  protected async _fillPartialTeleportParams(
```
Comment (godzillaba) [inline src/lib/assetBridger/l1l3Bridger.ts:1023]: this function is pretty complicated

**Comment by godzillaba** [comment]
would it be better for Erc20L1L3Bridger to have a private constructor, and instead have an `async create ...` function? that way it could check that the l1feetoken exists and is valid before you can actually use the bridger. It will also remove the need for an extra provider for some functions that need to get the l1 fee token.

File: src/lib/assetBridger/l1l3Bridger.ts [hunk 4]
```diff
@@ -0,0 +1,1419 @@
...
+   *
+   * If payment is skipped, the teleportation will not be completed until the L2 to L3 retryable is manually redeemed
+   *
+   * Defaults to false
+   */
+  skipFeeToken?: boolean
+  /**
+   * Optional recipient on L3, defaults to signer's address
+   */
+  to?: string
```
[the remark this answers is not in the document]
Comment (godzillaba) [reply src/lib/assetBridger/l1l3Bridger.ts:170]: 263b894091e756dac17b1d21c282c4b063035756

File: src/lib/assetBridger/l1l3Bridger.ts [hunk 5]
```diff
@@ -0,0 +1,1419 @@
...
+   */
+  l2Provider: Provider
+  /**
+   * L3 provider
+   */
+  l3Provider: Provider
+  /**
+   * Optional recipient on L3, defaults to signer's address
+   */
+  to?: string
```
[the remark this answers is not in the document]
Comment (godzillaba) [reply src/lib/assetBridger/l1l3Bridger.ts:226]: 263b894091e756dac17b1d21c282c4b063035756

File: src/lib/assetBridger/l1l3Bridger.ts [hunk 6]
```diff
@@ -0,0 +1,1419 @@
...
+/**
+ * Base functionality for L1 to L3 bridging.
+ */
+class BaseL1L3Bridger {
+  public readonly l1Network: L1Network
+  public readonly l2Network: L2Network
+  public readonly l3Network: L2Network
+
+  public readonly defaultGasPricePercentIncrease: BigNumber =
+    BigNumber.from(200)
```
[the remark this answers is not in the document]
Comment (godzillaba) [reply src/lib/assetBridger/l1l3Bridger.ts:265]: 7718d79bd5b2e2a67a0171ea46ece1134d7a846b

File: src/lib/assetBridger/l1l3Bridger.ts [hunk 7]
```diff
@@ -0,0 +1,1419 @@
...
+
+export enum TeleportationType {
+  /**
+   * Teleporting to an ETH fee L3
+   */
+  Standard,
+  /**
+   * Teleporting the fee token to a custom fee L3
+   */
+  OnlyFeeToken,
```
[the remark this answers is not in the document]
Comment (godzillaba) [reply src/lib/assetBridger/l1l3Bridger.ts:67]: 1c8b45b0a4eeca14eb2cfdb8c0862cb1b57ea493

File: src/lib/assetBridger/l1l3Bridger.ts [hunk 8]
```diff
@@ -0,0 +1,1419 @@
...
+  | { txHash: string }
+  | { tx: L1ContractCallTransaction }
+  | { txReceipt: L1ContractCallTransactionReceipt }
+)
+
+export type Erc20DepositMessages = {
+  /**
+   * L1 to L2 token bridge message
+   */
+  l1l2TokenBridge: L1ToL2MessageReader
```
[the remark this answers is not in the document]
Comment (godzillaba) [reply src/lib/assetBridger/l1l3Bridger.ts:191]: 1b99113cc0695c1073db142a6bdd900e171afac2

File: src/lib/assetBridger/l1l3Bridger.ts [hunk 9]
```diff
@@ -0,0 +1,1419 @@
...
+    }
+
+    this.teleporterAddresses = this.l2Network.teleporterAddresses
+  }
+
+  /**
+   * If the L3 network uses a custom fee token, return the address of that token on L1.
+   * If the fee token is not available on L1 or the L3 network uses ETH for fees, return the zero address.
+   */
+  public async l1FeeTokenAddress(
```
[the remark this answers is not in the document]
Comment (godzillaba) [reply src/lib/assetBridger/l1l3Bridger.ts:382]: e7f07b16849ca9d239ae6c3988442002f763e80c

**Review by spsjvc — APPROVED** [review]

**Pull request merged**